File: server/routers/canvas.py
from fastapi import APIRouter, Depends, HTTPException, Request
from services.chat_service import handle_chat
from services.db_service import db_service
from services.canvas_cover_service import (
    schedule_canvas_cover_generation,
    schedule_missing_canvas_covers,
    extract_prompt_from_messages,
    build_canvas_name_from_prompt,
    schedule_canvas_name_update,
    schedule_missing_canvas_names,
    DEFAULT_CANVAS_NAMES,
)
from routers.auth_router import get_current_user_id
import asyncio
import json

router = APIRouter(prefix="/api/canvas")


@router.get("/list")
async def list_canvases(user_id: str = Depends(get_current_user_id)):
    canvases = await db_service.list_canvases(user_id)
    schedule_missing_canvas_covers(canvases)
    schedule_missing_canvas_names(canvases)
    return canvases


@router.post("/create")
async def create_canvas(
    request: Request, user_id: str = Depends(get_current_user_id)
):
    data = await request.json()
    id = data.get("canvas_id")
    session_id = data.get("session_id")
    messages = data.get("messages", [])
    prompt = extract_prompt_from_messages(messages)
    fallback_name = data.get("name") or "未命名"
    name = build_canvas_name_from_prompt(prompt) if prompt else fallback_name
    data["user_id"] = user_id
    await db_service.create_canvas(id, name, session_id, user_id=user_id)
    await db_service.prune_canvases(user_id)
    asyncio.create_task(handle_chat(data))
    schedule_canvas_cover_generation(id, messages)
    if name in DEFAULT_CANVAS_NAMES:
        schedule_canvas_name_update(id, messages, session_id)
    return {"id": id}


@router.get("/{id}")
async def get_canvas(id: str, user_id: str = Depends(get_current_user_id)):
    canvas = await db_service.get_canvas_data(id, user_id=user_id)
    if canvas is None:
        raise HTTPException(status_code=404, detail="项目不存在或无权访问")
    return canvas


@router.post("/{id}/save")
async def save_canvas(
    id: str, request: Request, user_id: str = Depends(get_current_user_id)
):
    payload = await request.json()
    data_str = json.dumps(payload["data"])
    ok = await db_service.save_canvas_data(
        id, data_str, payload["thumbnail"], user_id=user_id
    )
    if not ok:
        raise HTTPException(status_code=404, detail="项目不存在或无权访问")
    return {"id": id}


@router.post("/{id}/rename")
async def rename_canvas(
    id: str, request: Request, user_id: str = Depends(get_current_user_id)
):
    data = await request.json()
    name = data.get("name")
    ok = await db_service.rename_canvas(id, name, user_id=user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="项目不存在或无权访问")
    return {"id": id}


@router.delete("/{id}/delete")
async def delete_canvas(id: str, user_id: str = Depends(get_current_user_id)):
    ok = await db_service.delete_canvas(id, user_id=user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="项目不存在或无权访问")
    return {"id": id}

File: react/src/api/canvas.ts
import { CanvasData, Message, Session } from '@/types/types'
import { ToolInfo } from '@/api/model'
import { authenticatedFetch } from './auth'

export type ListCanvasesResponse = {
  id: string
  name: string
  description?: string
  thumbnail?: string
  created_at: string
  updated_at?: string
  session_id?: string
}

function errorDetail(data: unknown, fallback: string): string {
  if (data && typeof data === 'object' && 'detail' in data) {
    const detail = (data as { detail: unknown }).detail
    if (typeof detail === 'string') return detail
  }
  return fallback
}

export async function listCanvases(): Promise<ListCanvasesResponse[]> {
  const response = await authenticatedFetch('/api/canvas/list')
  const data = await response.json().catch(() => ([]))
  if (!response.ok) {
    throw new Error(errorDetail(data, 'Failed to load projects'))
  }
  return data
}

export async function createCanvas(data: {
  name: string
  canvas_id: string
  messages: Message[]
  session_id: string
  text_model: {
    provider: string
    model: string
    url: string
  }
  tool_list: ToolInfo[]

  system_prompt: string
}): Promise<{ id: string }> {
  const response = await authenticatedFetch('/api/canvas/create', {
    method: 'POST',
    body: JSON.stringify(data),
  })
  const result = await response.json().catch(() => ({}))
  if (!response.ok) {
    throw new Error(errorDetail(result, 'Failed to create project'))
  }
  return result
}

export async function getCanvas(
  id: string
): Promise<{ data: CanvasData; name: string; sessions: Session[] }> {
  const response = await authenticatedFetch(`/api/canvas/${id}`)
  const data = await response.json().catch(() => ({}))
  if (!response.ok) {
    throw new Error(errorDetail(data, 'Failed to load project'))
  }
  return data
}

export async function saveCanvas(
  id: string,
  payload: {
    data: CanvasData
    thumbnail: string
  }
): Promise<void> {
  const response = await authenticatedFetch(`/api/canvas/${id}/save`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
  if (!response.ok) {
    const data = await response.json().catch(() => ({}))
    throw new Error(errorDetail(data, 'Failed to save project'))
  }
}

export async function renameCanvas(id: string, name: string): Promise<void> {
  const response = await authenticatedFetch(`/api/canvas/${id}/rename`, {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
  if (!response.ok) {
    const data = await response.json().catch(() => ({}))
    throw new Error(errorDetail(data, 'Failed to rename project'))
  }
}

export async function deleteCanvas(id: string): Promise<void> {
  const response = await authenticatedFetch(`/api/canvas/${id}/delete`, {
    method: 'DELETE',
  })
  if (!response.ok) {
    const data = await response.json().catch(() => ({}))
    throw new Error(errorDetail(data, 'Failed to delete project'))
  }
}
